File: src/app/state.ts
import { applyEvents } from "../domain/projector";
import { createEmptyState } from "../domain/state";
import { readConfig } from "../store/config";
import { readEvents } from "../store/events";
import { loadLatestSnapshot, writeSnapshot } from "../store/snapshots";
import { readStateCache, writeStateCache } from "../store/state";
import type { EventRecord, Snapshot, State } from "../types";

export interface LoadedState {
  state: State;
  allEvents: EventRecord[];
  warning?: string;
  snapshot: Snapshot | null;
}

export async function loadProjectedState(repoRoot: string): Promise<LoadedState> {
  const { events, warning } = await readEvents(repoRoot);
  const fromCache = await readStateCache(repoRoot);
  if (fromCache && fromCache.applied_events <= events.length) {
    const offset = fromCache.applied_events ?? 0;
    if (offset === events.length) {
      return {
        state: fromCache,
        allEvents: events,
        warning,
        snapshot: null,
      };
    }

    const state = applyEvents(fromCache, events.slice(offset));
    state.applied_events = events.length;
    return {
      state,
      allEvents: events,
      warning,
      snapshot: null,
    };
  }

  const snapshot = await loadLatestSnapshot(repoRoot);
  const base = snapshot ? snapshot.state : createEmptyState();
  const startOffset = snapshot ? Math.min(snapshot.event_count, events.length) : 0;
  const projected = applyEvents(base, events.slice(startOffset));
  projected.applied_events = events.length;

  return {
    state: projected,
    allEvents: events,
    warning,
    snapshot,
  };
}

export async function persistProjection(
  repoRoot: string,
  state: State,
  eventCount: number,
): Promise<void> {
  state.applied_events = eventCount;
  await writeStateCache(repoRoot, state);

  const config = await readConfig(repoRoot);
  if (config.snapshot_every <= 0) {
    return;
  }

  if (eventCount > 0 && eventCount % config.snapshot_every === 0) {
    await writeSnapshot(repoRoot, {
      taken_at: new Date().toISOString(),
      event_count: eventCount,
      state,
    });
  }
}

File: src-rust/src/app/state.rs
use crate::domain::projector::apply_events;
use crate::domain::state::create_empty_state;
use crate::errors::TsqError;
use crate::store::config::read_config;
use crate::store::events::read_events;
use crate::store::snapshots::{load_latest_snapshot_with_warning, write_snapshot};
use crate::store::state::{read_state_cache, write_state_cache};
use crate::types::{EventRecord, Snapshot, State};
use chrono::{SecondsFormat, Utc};
use std::path::Path;

pub struct LoadedState {
    pub state: State,
    pub all_events: Vec<EventRecord>,
    pub warning: Option<String>,
    pub snapshot: Option<Snapshot>,
}

pub fn load_projected_state(repo_root: impl AsRef<Path>) -> Result<LoadedState, TsqError> {
    let read = read_events(&repo_root)?;
    let events = read.events;
    let event_warning = read.warning;

    if let Some(from_cache) = read_state_cache(&repo_root)?
        && from_cache.applied_events <= events.len()
    {
        let offset = from_cache.applied_events;
        if offset == events.len() {
            return Ok(LoadedState {
                state: from_cache,
                all_events: events,
                warning: event_warning,
                snapshot: None,
            });
        }
        let mut state = apply_events(&from_cache, &events[offset..])?;
        state.applied_events = events.len();
        return Ok(LoadedState {
            state,
            all_events: events,
            warning: event_warning,
            snapshot: None,
        });
    }

    let loaded = load_latest_snapshot_with_warning(&repo_root)?;
    let base = loaded
        .snapshot
        .as_ref()
        .map(|snapshot| snapshot.state.clone())
        .unwrap_or_else(create_empty_state);
    let start_offset = loaded
        .snapshot
        .as_ref()
        .map(|snapshot| std::cmp::min(snapshot.event_count, events.len()))
        .unwrap_or(0);
    let mut projected = apply_events(&base, &events[start_offset..])?;
    projected.applied_events = events.len();

    Ok(LoadedState {
        state: projected,
        all_events: events,
        warning: combine_warnings(event_warning, loaded.warning),
        snapshot: loaded.snapshot,
    })
}

pub fn persist_projection(
    repo_root: impl AsRef<Path>,
    state: &mut State,
    event_count: usize,
    now: Option<&dyn Fn() -> String>,
) -> Result<(), TsqError> {
    state.applied_events = event_count;
    write_state_cache(&repo_root, state)?;

    let config = read_config(&repo_root)?;
    if config.snapshot_every == 0 {
        return Ok(());
    }

    if event_count > 0 && event_count.is_multiple_of(config.snapshot_every) {
        let taken_at = match now {
            Some(clock) => clock(),
            None => Utc::now().to_rfc3339_opts(SecondsFormat::Millis, true),
        };
        let snapshot = Snapshot {
            taken_at,
            event_count,
            state: state.clone(),
        };
        write_snapshot(&repo_root, &snapshot)?;
    }

    Ok(())
}

fn combine_warnings(warnings: Option<String>, other: Option<String>) -> Option<String> {
    let mut combined = Vec::new();
    if let Some(value) = warnings
        && !value.is_empty()
    {
        combined.push(value);
    }
    if let Some(value) = other
        && !value.is_empty()
    {
        combined.push(value);
    }
    if combined.is_empty() {
        None
    } else {
        Some(combined.join(" | "))
    }
}
